pkg/login/social/gitlab_oauth_test.go

```diff
@@ -173,3 +179,207 @@ func TestSocialGitlab_UserInfo(t *testing.T) {
 		})
 	}
 }
+
+type testCase struct {
+	name           string
+	payload        string
+	config         *oauth2.Config
+	wantUser       *userData
+	wantErrMessage string
+}
+
+func TestSocialGitlab_extractFromToken(t *testing.T) {
+	// Create a test server that returns a dummy ID token and user info
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case "/oauth/token":
+			// Return a dummy access token
+			_ = json.NewEncoder(w).Encode(map[string]interface{}{
+				"access_token": "dummy_access_token",
+				"token_type":   "Bearer",
+			})
+		case "/oauth/userinfo":
+			// Return a dummy user info
+			_ = json.NewEncoder(w).Encode(userInfoResponse{
+				Sub:           "12345678",
+				EmailVerified: true,
+				Groups:        []string{"admins", "editors", "viewers"},
+			})
+		default:
+			http.Error(w, "not found", http.StatusNotFound)
+		}
+	}))
+	defer ts.Close()
+
+	testCases := []testCase{
+		{
+			name: "successful extraction",
+			payload: `{
+                "iss": "https://gitlab.com",
+                "sub": "12345678",
+                "aud": "d77db857f4696c5c5ff6cee64f3ed26e709aac8f1c644dc4b9d5fd64f825d583",
+                "exp": 1686124040,
+                "iat": 1686123920,
+                "auth_time": 1686119303,
+                "sub_legacy": "b4359d63eaf90d4b1f3d71d291353b75a676bf73fdf734d4ff009eca5c69bb70",
+                "name": "John Doe",
+                "nickname": "johndoe",
+                "preferred_username": "johndoe",
+                "email": "johndoe@example.com",
+                "email_verified": true,
+                "profile": "https://gitlab.com/johndoe",
+                "picture": "https://gitlab.com/uploads/-/system/user/avatar/1234567/avatar.png",
+                "groups_direct": [
+                  "admins"
+                ]
```

Comment:
nit: could you only specify whether the email has or has not been verified for each test instead of specifying the full payload? Just to make them more readable.